File: TestApp/TestAppIOS/TestAppIOS/JSMessageHandler.swift
//
//  JSMessageHandler.swift
//  TestAppIOS
//
//  Created by Gary Griswold on 5/10/18.
//  Copyright © 2018 ShortSands. All rights reserved.
//
//

import Foundation
import UIKit
import WebKit
import Utility
import AWS
import AudioPlayer
import VideoPlayer

public class JSMessageHandler : NSObject, WKScriptMessageHandler {
    
    let controller: ViewController
    private var videoViewPlayer: VideoViewPlayer?

    init(controller: ViewController) {
        self.controller = controller
        super.init()
    }
    deinit {
        print("****** Deinitialize JSMessageHandler ******")
    }

    /**
     * var aMessage = {'command':'hello', data:[5,6,7,8,9]}
     * window.webkit.messageHandlers.callNative.postMessage(aMessage)
     */
    public func userContentController(_ userContentController: WKUserContentController, didReceive: WKScriptMessage) {
        print("CALL FROM JS: \(didReceive.body)")
        if didReceive.body is Dictionary<String, Any> {
            let request = (didReceive.body as? Dictionary<String, Any>)!
            let callbackId = request["callbackId"] as? String ?? "notString"
            let plugin = request["plugin"] as? String ?? "notString"
            let method = request["method"] as? String ?? "notString"
            let parameters = request["parameters"] as? [Any] ?? []
            if (plugin == "Utility") {
                utilityPlugin(callbackId: callbackId, method: "Utility." + method, parameters: parameters)
            } else if plugin == "Sqlite" {
                sqlitePlugin(callbackId: callbackId, method: "Sqlite." + method, parameters: parameters)
            } else if plugin == "AWS" {
                awsPlugin(callbackId: callbackId, method: "AWS." + method, parameters: parameters)
            } else if plugin == "AudioPlayer" {
                audioPlayerPlugin(callbackId: callbackId, method: "AudioPlayer." + method, parameters: parameters)
            } else if plugin == "VideoPlayer" {
                videoPlayerPlugin(callbackId: callbackId, method: "VideoPlayer." + method, parameters: parameters)
            } else {
                jsError(callbackId: callbackId, method: method, error: "Unknown plugin")
            }
        } else {
            print("Message set to callNative, must be a dictionary")
            // I think there is no other error response possible here.
        }
    }
    
    private func utilityPlugin(callbackId: String, method: String, parameters: [Any]) {
        
        if method == "Utility.locale" {
            let locale = Locale.current
            let localeStr = locale.identifier
            let language = locale.languageCode
            let script = locale.scriptCode
            let country = locale.regionCode
            print("locale \(localeStr)")
            let result = [localeStr, language, script, country]
            jsSuccess(callbackId: callbackId, plugin: "Utility", method: method, response: result)
            
        } else if method == "Utility.platform" {
            jsSuccess(callbackId: callbackId, response: "iOS")
            
        } else if method == "Utility.modelType" {
            jsSuccess(callbackId: callbackId, response: UIDevice.current.model)
            
        } else if method == "Utility.modelName" {
            let modelName = DeviceSettings.modelName()
            jsSuccess(callbackId: callbackId, response: modelName)
            
        } else if method == "Utility.deviceSize" {
            let deviceSize = DeviceSettings.deviceSize()
            jsSuccess(callbackId: callbackId, response: deviceSize)
            
        } else if method == "Utility.hideKeyboard" {
            let hidden = self.controller.webview.endEditing(true)
            jsSuccess(callbackId: callbackId, response: hidden)
            
        } else {
            jsError(callbackId: callbackId, method: method, error: "unknown method")
        }
    }
    
    private func sqlitePlugin(callbackId: String, method: String, parameters: [Any]) {
        
        if method == "Sqlite.openDB" {
            if parameters.count == 2 {
                let dbname = parameters[0] as? String ?? "notString"
                let isCopyDatabase = parameters[1] as? Bool ?? true
                do {
                    try Sqlite3.openDB(dbname: dbname, copyIfAbsent: isCopyDatabase)
                    jsSuccess(callbackId: callbackId)
                } catch let err {
                    jsError(callbackId: callbackId, method: method, error: err.localizedDescription)
                }
            } else {
                jsError(callbackId: callbackId, method: method, error: "Must have two parameters")
            }
            
        } else if method == "Sqlite.queryJS" {
            if parameters.count == 3 {
                let dbname = parameters[0] as? String ?? "notString"
                let statement = parameters[1] as? String ?? "notString"
                let values = parameters[2] as? [Any] ?? []
                do {
                    let db = try Sqlite3.findDB(dbname: dbname)
                    let result: [Dictionary<String,Any?>] = try db.queryV0(sql: statement, values: values)
                    jsSuccess(callbackId: callbackId, plugin: "Sqlite", method: method, response: result)
                } catch let err {
                    jsError(callbackId: callbackId, method: method, error: err.localizedDescription, defaultVal: [])
                }
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have three parameters", defaultVal: [])
            }
            
        } else if method == "Sqlite.executeJS" {
            if parameters.count == 3 {
                let dbname = parameters[0] as? String ?? "notString"
                let statement = parameters[1] as? String ?? "notString"
                let values = parameters[2] as? [Any] ?? []
                do {
                    let db = try Sqlite3.findDB(dbname: dbname)
                    let result: Int = try db.executeV1(sql: statement, values: values)
                    jsSuccess(callbackId: callbackId, response: result)
                } catch let err {
                    jsError(callbackId: callbackId, method: method, error: err.localizedDescription, defaultVal: 0)
                }
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have three parameters", defaultVal: 0)
            }
            
        } else if method == "Sqlite.bulkExecuteJS" {
            if parameters.count == 3 {
                let dbname = parameters[0] as? String ?? "notString"
                let statement = parameters[1] as? String ?? "notString"
                let values = parameters[2] as? [[Any]] ?? [[]]
                do {
                    let db = try Sqlite3.findDB(dbname: dbname)
                    let result: Int = try db.bulkExecuteV1(sql: statement, values: values)
                    jsSuccess(callbackId: callbackId, response: result)
                } catch let err {
                    jsError(callbackId: callbackId, method: method, error: err.localizedDescription, defaultVal: 0)
                }
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have three parameters", defaultVal: 0)
            }
            
        } else if method == "Sqlite.closeDB" {
            if parameters.count == 1 {
                let dbname = parameters[0] as? String ?? "notString"
                Sqlite3.closeDB(dbname: dbname)
                jsSuccess(callbackId: callbackId)
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have one parameter")
            }

        } else if method == "Sqlite.listDB" {
            do {
                let result: [String] = try Sqlite3.listDB()
                jsSuccess(callbackId: callbackId, plugin: "Sqlite", method: method, response: result)
            } catch let err {
                jsError(callbackId: callbackId, method: method, error: err.localizedDescription, defaultVal: [])
            }
            
        } else if method == "Sqlite.deleteDB" {
            if parameters.count == 1 {
                let dbname = parameters[0] as? String ?? "notString"
                do {
                    try Sqlite3.deleteDB(dbname: dbname)
                    jsSuccess(callbackId: callbackId)
                } catch let err {
                    jsError(callbackId: callbackId, method: method, error: err.localizedDescription)
                }
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have one parameter")
            }
        } else {
            jsError(callbackId: callbackId, method: method, error: "unknown method")
        }
    }
    
    private func awsPlugin(callbackId: String, method: String, parameters: [Any]) {

        if method == "AWS.downloadZipFile" {
            if parameters.count == 4 {
                let regionType = parameters[0] as? String ?? "notString"
                let s3Bucket = parameters[1] as? String ?? "notString"
                let s3Key = parameters[2] as? String ?? "notString"
                let filePath = parameters[3] as? String ?? "notString"
                let fileURL = URL(fileURLWithPath: NSHomeDirectory() + filePath)
                var s3: AwsS3? = nil
                if regionType == "SS" {
                    s3 = AwsS3Manager.findSS()
                } else if regionType == "DBP" {
                    s3 = AwsS3Manager.findDbp()
                } else if regionType == "TEST" {
                    s3 = AwsS3Manager.findTest()
                } else {
                    jsError(callbackId: callbackId, method: method, error: "Region must be SS, DBP, or TEST")
                }
                if let awsS3 = s3 {
                    awsS3.downloadZipFile(s3Bucket: s3Bucket, s3Key: s3Key, filePath: fileURL, view: nil,
                                          complete: { err in
                        if let err1 = err {
                            self.jsError(callbackId: callbackId, method: method, error: err1.localizedDescription)
                        } else {
                            self.jsSuccess(callbackId: callbackId)
                        }
                    })
                }
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have three parameters")
            }
        } else {
            jsError(callbackId: callbackId, method: method, error: "unknown method")
        }
    }
    
    private func audioPlayerPlugin(callbackId: String, method: String, parameters: [Any]) {
        
        if method == "AudioPlayer.findAudioVersion" {
            if parameters.count == 2 {
                let audioController = AudioBibleController.shared
                audioController.findAudioVersion(
                    version: parameters[0] as? String ?? "notString",
                    silLang: parameters[1] as? String ?? "notString",
                    complete: { bookIdList in
                        self.jsSuccess(callbackId: callbackId, response: bookIdList)
                })
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have two parameters", defaultVal: "")
            }

        } else if method == "AudioPlayer.isPlaying" {
            let audioController = AudioBibleController.shared
            let result: String = (audioController.isPlaying()) ? "T" : "F"
            jsSuccess(callbackId: callbackId, response: result)
            
        } else if method == "AudioPlayer.present" {
            if parameters.count == 2 {
                let audioController = AudioBibleController.shared
                audioController.present(view: controller.webview,
                                        book: parameters[0] as? String ?? "notString",
                                        chapterNum: parameters[1] as? Int ?? 1,
                                        complete: { error in
                                            // No error is actually being returned
                                            if let err = error {
                                                self.jsError(callbackId: callbackId, method: method,
                                                             error: err.localizedDescription)
                                            } else {
                                                self.jsSuccess(callbackId: callbackId)
                                            }
                })
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have two parameters")
            }
        
        } else if method == "AudioPlayer.stop" {
            let audioController = AudioBibleController.shared
            audioController.stop()
            jsSuccess(callbackId: callbackId)
            
        } else {
            jsError(callbackId: callbackId, method: method, error: "unknown method")
        }
    }
    
    private func videoPlayerPlugin(callbackId: String, method: String, parameters: [Any]) {
        
        if method == "VideoPlayer.showVideo" {
            if parameters.count == 5 {
                let player = VideoViewPlayer(mediaSource: parameters[0] as? String ?? "notString",
                                             videoId: parameters[1] as? String ?? "notString",
                                             languageId: parameters[2] as? String ?? "notString",
                                             silLang: parameters[3] as? String ?? "notString",
                                             videoUrl: parameters[4] as? String ?? "notString")
                self.videoViewPlayer = player
                player.begin(complete: { error in
                    if let err = error {
                        self.jsError(callbackId: callbackId, method: method, error: err.localizedDescription)
                    } else {
                        self.jsSuccess(callbackId: callbackId)
                    }
                })
                self.controller.present(player.controller, animated: true)
            } else {
                jsError(callbackId: callbackId, method: method, error: "must have five parameters")
            }
        } else {
            jsError(callbackId: callbackId, method: method, error: "unknown method")
        }
    }
    
    /**
    * Success Callbacks
    */
    private func jsSuccess(callbackId: String) {
        jsCallback(callbackId: callbackId, json: false, error: nil, response: "null")
    }
    
    private func jsSuccess(callbackId: String, response: Int) {
        jsCallback(callbackId: callbackId, json: false, error: nil, response: String(response))
    }
    
    private func jsSuccess(callbackId: String, response: Bool) {
        jsCallback(callbackId: callbackId, json: false, error: nil, response: String(response))
    }
    
    private func jsSuccess(callbackId: String, response: String?) {
        let result: String = (response != nil) ? "'" + response! + "'" : "null"
        jsCallback(callbackId: callbackId, json: false, error: nil, response: result)
    }
    
    private func jsSuccess(callbackId: String, plugin: String, method: String, response: [Any?]) {
        do {
            let message: Data = try JSONSerialization.data(withJSONObject: response)
            let result = "'" + String(data: message, encoding: String.Encoding.utf8)! + "'"
            jsCallback(callbackId: callbackId, json: true, error: nil, response: result)
        } catch let jsonErr {
            let error = logError(method: method, message: jsonErr.localizedDescription)
            jsCallback(callbackId: callbackId, json: false, error: error, response: "null")
        }
    }
    
    /**
    * Error Callbacks
    */
    private func jsError(callbackId: String, method: String, error: String) {
        let err = logError(method: method, message: error)
        jsCallback(callbackId: callbackId, json: false, error: err, response: "null")
    }
    
    private func jsError(callbackId: String, method: String, error: String, defaultVal: Bool) {
        let err = logError(method: method, message: error)
        jsCallback(callbackId: callbackId, json: false, error: err, response: String(defaultVal))
    }
    
    private func jsError(callbackId: String, method: String, error: String, defaultVal: Int) {
        let err = logError(method: method, message: error)
        jsCallback(callbackId: callbackId, json: false, error: err, response: String(defaultVal))
    }
    
    private func jsError(callbackId: String, method: String, error: String, defaultVal: String) {
        let err = logError(method: method, message: error)
        let response = "'" + defaultVal + "'"
        jsCallback(callbackId: callbackId, json: false, error: err, response: response)
    }
    
    private func jsError(callbackId: String, method: String, error: String, defaultVal: [Any?]) {
        do {
            let message: Data = try JSONSerialization.data(withJSONObject: defaultVal)
            let result = "'" + String(data: message, encoding: String.Encoding.utf8)! + "'"
            let err = logError(method: method, message: error)
            jsCallback(callbackId: callbackId, json: true, error: err, response: result)
        } catch let jsonErr {
            let error = logError(method: method, message: jsonErr.localizedDescription)
            jsCallback(callbackId: callbackId, json: false, error: error, response: "null")
        }
    }
    
    private func jsCallback(callbackId: String, json: Bool, error: String?, response: Any) {
        let isJson: Int = (json) ? 1 : 0
        let err = (error != nil) ? "'" + error! + "'" : "null"
        let message = "handleNative('\(callbackId)', \(isJson), \(err), \(response));"
        print("RETURN TO JS: \(message)")
        self.controller.webview.evaluateJavaScript(message, completionHandler: { data, error in
            if let err = error {
                print("jsCallbackError \(err)")
            }
            if let resp = data {
                print("jsCallback has unexpected response \(resp)")
            }
        })
    }
    
    private func logError(method: String, message: String) -> String {
        let error = "PLUGIN ERROR: \(method): \(message)"
        print(error)
        return error
    }
}

File: SafeBible/SafeBible_Android/app/src/main/java/com/shortsands/safebible/JSMessageHandler.java
package com.shortsands.safebible;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Bundle;
import android.provider.Settings;
import android.view.inputmethod.InputMethodManager;
import android.view.View;
import android.util.Log;
import android.webkit.JavascriptInterface;
import android.webkit.ValueCallback;
import com.shortsands.audioplayer.AudioBibleController;
import com.shortsands.aws.AwsS3;
import com.shortsands.aws.AwsS3Manager;
import com.shortsands.aws.CompletionHandler;
import com.shortsands.aws.DownloadZipFileListener;
import com.shortsands.utility.Sqlite3;
import com.shortsands.videoplayer.VideoActivity;
import java.io.File;
import java.util.ArrayList;
import java.util.Locale;
import org.json.JSONArray;

/**
 * Created by garygriswold on 5/10/18.
 */

public class JSMessageHandler {

    private static final String TAG = "JSMessageHandler";

    private MainActivity activity;

    public JSMessageHandler(MainActivity activity) {
        this.activity = activity;
    }

    /**
     *
     */
    @JavascriptInterface
    public void jsHandler(String callbackId, String plugin, String method, String parameterStr) {
        try {
            if (plugin.equals("console")) {
                Log.d("Javascript", parameterStr);
            } else {
                Log.d(TAG, "CALL FROM JS: " + plugin + "." + method + " " + parameterStr);
                JSONArray parameters = new JSONArray(parameterStr);
                if (plugin.equals("Utility")) {
                    utilityPlugin(callbackId, "Utility." + method, parameters);
                } else if (plugin.equals("Sqlite")) {
                    sqlitePlugin(callbackId, "Sqlite." + method, parameters);
                } else if (plugin.equals("AWS")) {
                    awsPlugin(callbackId, "AWS." + method, parameters);
                } else if (plugin.equals("AudioPlayer")) {
                    audioPlayerPlugin(callbackId, "AudioPlayer." + method, parameters);
                } else if (plugin.equals("VideoPlayer")) {
                    videoPlayerPlugin(callbackId, "VideoPlayer." + method, parameters);
                } else {
                    jsError(callbackId, method, "Unknown plugin");
                }
            }
        } catch(Exception err) {
            jsError(callbackId, plugin + "." + method, err.toString());
        }
    }

    private void utilityPlugin(String callbackId, String method, JSONArray parameters) {
        if (method.equals("Utility.locale")) {
            JSONArray result = new JSONArray();
            Locale locale = Locale.getDefault();
            result.put(locale.toString().replace("_", "-"));
            result.put(locale.getLanguage());
            result.put(locale.getScript());
            result.put(locale.getCountry());
            Log.d(TAG, "locale " + locale.toString());
            jsSuccess(callbackId, result);

        } else if (method.equals("Utility.platform")) {
            jsSuccess(callbackId, "Android");

        } else if (method.equals("Utility.modelType")) {
            jsSuccess(callbackId, android.os.Build.BRAND);

        } else if (method.equals("Utility.modelName")) {
            jsSuccess(callbackId, android.os.Build.MODEL);

            //} else if (method.equals("Utility.deviceSize")) {
            //    String deviceSize = DeviceSettings.deviceSize();
            //    jsSuccess(callbackId, deviceSize);

        } else if (method.equals("Utility.hideKeyboard")) {
            try {
                InputMethodManager inputManager = (InputMethodManager) this.activity.getSystemService(Context.INPUT_METHOD_SERVICE);
                View v = this.activity.getCurrentFocus();
                if (v != null) {
                    inputManager.hideSoftInputFromWindow(v.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
                    jsSuccess(callbackId, true);
                } else {
                    jsError(callbackId, method, "No current focus");
                }
            } catch (Exception error) {
                jsError(callbackId, method, error.toString() + " on hideKeyboard");
            }

        } else if (method.equals("Utility.rateApp")) {
            // Not yet implemented
            jsSuccess(callbackId);

        } else {
            jsError(callbackId, method, "unknown method");
        }
    }

    private void sqlitePlugin(String callbackId, String method, JSONArray parameters) {

        if (method.equals("Sqlite.openDB")) {
            if (parameters != null && parameters.length() == 2) {
                try {
                    String dbname = parameters.getString(0);
                    boolean copyIfAbsent = parameters.getBoolean(1);
                    Sqlite3.openDB(this.activity, dbname, copyIfAbsent);
                    jsSuccess(callbackId);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString());
                }
            } else {
                jsError(callbackId, method, "Must have two parameters");
            }

        } else if (method.equals("Sqlite.queryJS")) {
            if (parameters != null && parameters.length() == 3) {
                try {
                    String dbname = parameters.getString(0);
                    String statement = parameters.getString(1);
                    JSONArray values = parameters.getJSONArray(2);
                    Sqlite3 db = Sqlite3.findDB(dbname);
                    JSONArray result = db.queryJS(statement, values);
                    jsSuccess(callbackId, result);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString(), new JSONArray());
                }
            } else {
                jsError(callbackId, method, "must have three parameters", new JSONArray());
            }

        } else if (method.equals("Sqlite.queryHTML")) {
            if (parameters != null && parameters.length() == 3) {
                try {
                    String dbname = parameters.getString(0);
                    String statement = parameters.getString(1);
                    JSONArray values = parameters.getJSONArray(2);
                    Sqlite3 db = Sqlite3.findDB(dbname);
                    String result1 = db.queryHTMLv0(statement, values);
                    String result2 = result1.replace("\r", "\\r");
                    String result3 = result2.replace("\n", "\\n");
                    jsSuccess(callbackId, result3);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString(), "");
                }
            } else {
                //let err = JSMessageError.mustHaveParameters(method: method, num: 3)
                jsError(callbackId, method, "must have three parameters", "");
            }

        } else if (method.equals("Sqlite.querySSIF")) {
            if (parameters != null && parameters.length() == 3) {
                try {
                    String dbname = parameters.getString(0);
                    String statement = parameters.getString(1);
                    JSONArray values = parameters.getJSONArray(2);
                    Sqlite3 db = Sqlite3.findDB(dbname);
                    String result = db.querySSIFv0(statement, values);
                    jsSuccess(callbackId, result);
                } catch(Exception err) {
                    jsError(callbackId, err.toString(), "");
                }
            } else {
                //let err = JSMessageError.mustHaveParameters(method: method, num: 3)
                jsError(callbackId, "must have three parameters", "");
            }

        } else if (method.equals("Sqlite.executeJS")) {
            if (parameters != null && parameters.length() == 3) {
                try {
                    String dbname = parameters.getString(0);
                    String statement = parameters.getString(1);
                    JSONArray values = parameters.getJSONArray(2);
                    Sqlite3 db = Sqlite3.findDB(dbname);
                    int result = db.executeJS(statement, values);
                    jsSuccess(callbackId, result);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString(), 0);
                }
            } else {
                jsError(callbackId, method, "must have three parameters", 0);
            }

        } else if (method.equals("Sqlite.bulkExecuteJS")) {
            if (parameters != null && parameters.length() == 3) {
                try {
                    String dbname = parameters.getString(0);
                    String statement = parameters.getString(1);
                    JSONArray values = parameters.getJSONArray(2);
                    Sqlite3 db = Sqlite3.findDB(dbname);
                    int result = db.bulkExecuteJS(statement, values);
                    jsSuccess(callbackId, result);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString(), 0);
                }
            } else {
                jsError(callbackId, method, "must have three parameters", 0);
            }

        } else if (method.equals("Sqlite.closeDB")) {
            if (parameters != null && parameters.length() == 1) {
                try {
                    String dbname = parameters.getString(0);
                    Sqlite3.closeDB(dbname);
                    jsSuccess(callbackId);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString());
                }
            } else {
                jsError(callbackId, method, "must have one parameter");
            }

        } else if (method.equals("Sqlite.listDB")) {
            try {
                ArrayList<String> files = Sqlite3.listDB(this.activity);
                JSONArray result = new JSONArray(files);
                jsSuccess(callbackId, result);
            } catch(Exception err) {
                jsError(callbackId, method, err.toString(), new JSONArray());
            }

        } else if (method.equals("Sqlite.deleteDB")) {
            if (parameters != null && parameters.length() == 1) {
                try {
                    String dbname = parameters.getString(0);
                    Sqlite3.deleteDB(this.activity, dbname);
                    jsSuccess(callbackId);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString());
                }
            } else {
                jsError(callbackId, method, "must have one parameter");
            }

        } else {
            jsError(callbackId, method, "unknown method");
        }
    }

    private void awsPlugin(String callbackId, String method, JSONArray parameters) {
        AwsS3Manager.initialize(this.activity);

        if (method.equals("AWS.downloadZipFile")) {
            if (parameters != null && parameters.length() == 4) {
                try {
                    String regionType = parameters.getString(0);
                    String s3Bucket = parameters.getString(1);
                    String s3Key = parameters.getString(2);
                    String filePath = parameters.getString(3);
                    File file = new File(this.activity.getDataDir(), filePath);
                    AwsS3 s3 = null;
                    if (regionType.equals("SS")) {
                        s3 = AwsS3Manager.findSS();
                    } else if (regionType.equals("DBP")) {
                        s3 = AwsS3Manager.findDbp();
                    } else if (regionType.equals("TEST")) {
                        s3 = AwsS3Manager.findTest();
                    } else {
                        jsError(callbackId, method, "Region must be SS, DBP, or TEST");
                    }
                    if (s3 != null) {
                        DownloadPluginZipFileListener listener = new DownloadPluginZipFileListener(this, callbackId, method);
                        listener.setActivity(this.activity); // presents ProgressCircle
                        s3.downloadZipFile(s3Bucket, s3Key, file, listener);
                    }
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString());
                }
            } else {
                jsError(callbackId, method, "must have three parameters");
            }

        } else {
            jsError(callbackId, method, "unknown method");
        }
    }

    class DownloadPluginZipFileListener extends DownloadZipFileListener {

        private JSMessageHandler jsMessageHandler;
        private String callbackId;
        private String method;

        private File unzipped = null;

        public DownloadPluginZipFileListener(JSMessageHandler jsMessageHandler, String callbackId, String method) {
            super();
            this.jsMessageHandler = jsMessageHandler;
            this.callbackId = callbackId;
            this.method = method;
        }

        @Override
        protected void onComplete(int id) {
            super.onComplete(id);
            this.jsMessageHandler.jsSuccess(this.callbackId);
        }

        @Override
        public void onError(int id, Exception error) {
            super.onError(id, error);
            this.jsMessageHandler.jsError(this.callbackId, this.method, error.toString() + " on " + file.getAbsolutePath());
        }
    }

    private void audioPlayerPlugin(String callbackId, String method, JSONArray parameters) {
        AwsS3Manager.initialize(this.activity);

        if (method.equals("AudioPlayer.findAudioVersion")) {
            if (parameters != null && parameters.length() == 2) {
                try {
                    String version = parameters.getString(0);
                    String silLang = parameters.getString(1);
                    AudioBibleController audioController = AudioBibleController.shared(this.activity);
                    String bookList = audioController.findAudioVersion(version, silLang);
                    jsSuccess(callbackId, bookList);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString());
                }
            } else {
                jsError(callbackId, method, "must have two parameters", "");
            }

        } else if (method.equals("AudioPlayer.isPlaying")) {
            AudioBibleController audioController = AudioBibleController.shared(this.activity);
            String result = (audioController.isPlaying()) ? "T" : "F";
            jsSuccess(callbackId, result);

        } else if (method.equals("AudioPlayer.present")) {
            if (parameters != null && parameters.length() == 2) {
                try {
                    String book = parameters.getString(0);
                    int chapter = parameters.getInt(1);
                    AudioBibleController audioController = AudioBibleController.shared(this.activity);
                    AudioPresentCompletion complete = new AudioPresentCompletion(callbackId, method);
                    audioController.present(this.activity.getWebview(), book, chapter, complete);
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString());
                }
            } else {
                jsError(callbackId, method, "must have two parameters");
            }

        } else if (method.equals("AudioPlayer.stop")) {
            AudioBibleController audioController = AudioBibleController.shared(this.activity);
            audioController.stop();
            jsSuccess(callbackId);

        } else {
            jsError(callbackId, method, "unknown method");
        }
    }

    class AudioPresentCompletion implements CompletionHandler {
        private String callbackId;
        private String method;

        public AudioPresentCompletion(String callbackId, String method) {
            this.callbackId = callbackId;
            this.method = method;
        }
        @Override
        public void completed(Object result) {
            jsSuccess(this.callbackId);
        }
        @Override
        public void failed(Throwable exception) {
            Log.d(TAG, "NextReadFile Failed " + exception.toString());
            jsError(this.callbackId, this.method, exception.toString());
        }
    }

    private void videoPlayerPlugin(final String callbackId, final String method, final JSONArray parameters) {
        AwsS3Manager.initialize(this.activity);

        if (method.equals("VideoPlayer.showVideo")) {
            if (parameters != null && parameters.length() == 5) {
                final Intent videoIntent = new Intent(activity.getApplicationContext(), VideoActivity.class);
                Bundle extras = new Bundle();
                try {
                    extras.putString("mediaSource", parameters.getString(0));
                    extras.putString("videoId", parameters.getString(1));
                    extras.putString("languageId", parameters.getString(2));
                    extras.putString("silLang", parameters.getString(3));
                    extras.putString("videoUrl", parameters.getString(4));
                    videoIntent.putExtras(extras);
                    this.activity.startVideoActivity(callbackId, method, videoIntent);
                    // See method this.activity.onActivityResult for response handling
                } catch(Exception err) {
                    jsError(callbackId, method, err.toString());
                }
            } else {
                jsError(callbackId, method, "must have five parameters");
            }

        } else {
            jsError(callbackId, method, "unknown method");
        }
    }

    /**
     * Success Callbacks
     */
    public void jsSuccess(String callbackId) {
        jsCallback(callbackId, false, null, "null");
    }

    private void jsSuccess(String callbackId, int response) {
        jsCallback(callbackId, false, null, String.valueOf(response));
    }

    private void jsSuccess(String callbackId, boolean response) {
        jsCallback(callbackId, false, null, String.valueOf(response));
    }

    private void jsSuccess(String callbackId, String response) {
        String result = (response != null) ? "'" + response.replace("'", "&apos;") + "'" : "null";
        jsCallback(callbackId, false, null, result);
    }

    private void jsSuccess(String callbackId, JSONArray response) {
        String result = "'" + response.toString() + "'";
        jsCallback(callbackId, true, null, result);
    }

    /**
     * Error Callbacks
     */
    public void jsError(String callbackId, String method, String error) {
        String err = logError(method, error);
        jsCallback(callbackId, false, err, "null");
    }

    private void jsError(String callbackId, String method, String error, boolean defaultVal) {
        String err = logError(method, error);
        jsCallback(callbackId, false, err, String.valueOf(defaultVal));
    }

    private void jsError(String callbackId, String method, String error, int defaultVal) {
        String err = logError(method, error);
        jsCallback(callbackId, false, err, String.valueOf(defaultVal));
    }

    private void jsError(String callbackId, String method, String error, String defaultVal) {
        String err = logError(method, error);
        //String response = "'" + defaultVal.replace("'", "\'") + "'";
        String response = "'" + defaultVal + "'";
        jsCallback(callbackId, false, err, response);
    }

    private void jsError(String callbackId, String method, String error, JSONArray defaultVal) {
        String err = logError(method, error);
        String response = "'" + defaultVal.toString() + "'";
        jsCallback(callbackId, true, err, response);
    }

    private void jsCallback(String callbackId, boolean json, String error, String response) {
        int isJson = (json) ? 1 : 0;
        String err = (error != null) ? "'" + error + "'" : "null";
        final String message = "handleNative('" + callbackId + "', " + isJson + ", " + err + ", " + response + ");";
        Log.d(TAG, "RETURN TO JS: " + message);
        this.activity.runOnUiThread(new Runnable() {
            public void run() {
                activity.getWebview().evaluateJavascript(message, new ValueCallback<String>() {
                    @Override
                    public void onReceiveValue(String str) {
                        if (!str.equals("null")) {
                            Log.d("jsCallbackError", str);
                        }
                    }
                });
            }
        });
    }

    private String logError(String method, String message) {
        String error = "PLUGIN ERROR: " + method + ": " + message;
        Log.e(TAG, error);
        return error;
    }
}

